lib/dat-worker-pool/runner.rb

```diff
@@ -26,22 +26,17 @@ def start
       @num_workers.times{ build_worker }
     end
 
-    # the workers should be told to shutdown before the queue because the queue
-    # shutdown will wake them up; a worker popping on a shutdown queue will
-    # always get `nil` back and will loop as fast as allowed until its shutdown
-    # flag is flipped, so shutting down the workers then the queue keeps them from
-    # looping as fast as possible; use an until loop instead of each to join all
-    # the workers, while we are joining a worker a different worker can shutdown
-    # and remove itself from the `@workers` array
-    def shutdown(timeout = nil)
+    # if any kind of standard error or the expected timeout error (assuming the
+    # workers take to long to shutdown) is raised - force shutdown, this ensures
+    # we shutdown as best as possible instead of letting ruby kill the threads
+    # when the process exits; non-timeout errors will be re-raised so they can
+    # be caught and handled (or shown when the process exits)
+    def shutdown(timeout = nil, &backtrace_proc)
       begin
-        OptionalTimeout.new(timeout) do
-          @workers.each(&:dwp_shutdown)
-          @queue.shutdown
-          @workers.first.dwp_join until @workers.empty?
-        end
-      rescue TimeoutError
-        force_shutdown(timeout, caller)
+        graceful_shutdown(timeout)
+      rescue StandardError, TimeoutInterruptError => exception
+        forced_shutdown(exception, timeout, &backtrace_proc)
+        raise exception if !exception.kind_of?(TimeoutInterruptError)
```

Comment:
@kellyredding - Really only `queue.shutdown` should/could ever raise a standard error since its possible for custom queues to have unsafe shutdown logic and since we rescue errors on `worker.join`. This is also better for any future changes in the case shutdown ever does more or `worker.shutdown` ever does more.